vm/src/slots.rs

```diff
@@ -35,25 +39,78 @@ impl Default for PyTpFlags {
     }
 }
 
+type GenericMethod = fn(PyObjectRef, PyFuncArgs, &VirtualMachine) -> PyResult;
+type DescrGetFunc =
+    fn(PyObjectRef, Option<PyObjectRef>, Option<PyObjectRef>, &VirtualMachine) -> PyResult;
+type HashFunc = fn(PyObjectRef, &VirtualMachine) -> PyResult<PyHash>;
+
 #[derive(Default)]
 pub struct PyClassSlots {
     pub flags: PyTpFlags,
     pub name: PyRwLock<Option<String>>, // tp_name, not class name
     pub new: Option<PyNativeFunc>,
-    pub call: Option<PyNativeFunc>,
-    pub descr_get: Option<PyDescrGetFunc>,
-    pub hash: Option<HashFunc>,
+    pub call: AtomicCell<Option<GenericMethod>>,
+    pub descr_get: AtomicCell<Option<DescrGetFunc>>,
+    pub hash: AtomicCell<Option<HashFunc>>,
 }
 
-type HashFunc = Box<py_dyn_fn!(dyn Fn(PyObjectRef, &VirtualMachine) -> PyResult<PyHash>)>;
-
 impl PyClassSlots {
     pub fn from_flags(flags: PyTpFlags) -> Self {
         Self {
             flags,
             ..Default::default()
         }
     }
+
+    #[inline]
+    pub(crate) fn update_slot_func(&self, name: &str) {
+        match name {
+            "__call__" => {
+                let func: GenericMethod = |zelf, args, vm| {
+                    let magic = get_class_magic(&zelf, "__call__");
+                    let magic = vm.call_if_get_descriptor(magic, zelf.clone())?;
+                    args.insert(zelf);
+                    vm.invoke(&magic, args)
+                } as _;
+                self.call.store(Some(func))
+            }
+            "__get__" => {
+                let func: DescrGetFunc = |zelf, obj, cls, vm| {
+                    let magic = get_class_magic(&zelf, "__get__");
+                    vm.invoke(
+                        &magic,
+                        vec![zelf, vm.unwrap_or_none(obj), vm.unwrap_or_none(cls)],
+                    )
+                } as _;
+                self.descr_get.store(Some(func))
+            }
+            "__hash__" => {
+                let func: HashFunc = |zelf, vm| {
+                    let magic = get_class_magic(&zelf, "__hash__");
+                    let hash_obj = vm.invoke(&magic, vec![zelf])?;
+                    match hash_obj.payload_if_subclass::<PyInt>(vm) {
+                        Some(py_int) => {
+                            Ok(rustpython_common::hash::hash_bigint(py_int.borrow_value()))
+                        }
+                        None => Err(vm
+                            .new_type_error("__hash__ method should return an integer".to_owned())),
+                    }
+                } as _;
+                self.hash.store(Some(func));
+            }
+            _ => (),
+        }
+    }
+}
+
+#[inline]
+fn get_class_magic(zelf: &PyObjectRef, name: &str) -> PyObjectRef {
+    zelf.get_class_attr(name).unwrap()
+
+    // TODO: we already looked up the matching class but lost the information here
+    // let cls = zelf.lease_class();
+    // let attrs = cls.attributes.read();
+    // attrs.get(name).unwrap().clone()
```

Comment:
we are double looking up the matching class. can slot refer its class? weak ref?

Reply:
Are you looking to sort of store the class that has the `__call__` method in the call slot, as if it's a closure, so that you can just do `class_call.attrs.get("__call__")`? not suggesting that we make it a closure, just trying to figure out what the optimization would be?

Reply:
Basically yes. We can find the class 2 ways.
1. In update_slot_func, the owner of `__call__` is always the type owning the slots. 
2. When finding Some() slot in first_in_mro, it is always the type owning `__call__`

Reply:
giving additional class parameter to slot funcs will work